policy-controller/src/main.rs

```diff
@@ -111,17 +121,70 @@ async fn main() -> Result<()> {
     // Run the gRPC server, serving results by looking up against the index handle.
     tokio::spawn(grpc(grpc_addr, cluster_networks, handle, drain_rx.clone()));
 
-    // Run the admission controller
     if let Some(bind_addr) = admission_addr {
-        let (listen_addr, serve) = warp::serve(admission::routes(client))
-            .tls()
-            .cert_path("/var/run/linkerd/tls/tls.crt")
-            .key_path("/var/run/linkerd/tls/tls.key")
-            .bind_with_graceful_shutdown(bind_addr, async move {
-                let _ = drain_rx.signaled().await;
-            });
+        let tcp = TcpListener::bind(&bind_addr).await?;
+        let listen_addr = tcp.local_addr()?;
+        // Prepare a long-running future stream to accept and serve clients.
+        let incoming_tls_stream = stream! {
+            loop {
+                let socket = match tcp.accept().await {
+                    Ok((socket, _)) => socket,
+                    Err(err) => {
+                        error!(%err, "Failed to accept connection");
+                        continue;
+                    }
+                };
+                let tls_cfg = {
+                    // Load public certificate.
+                    let certs = match load_certs("/var/run/linkerd/tls/tls.crt") {
+                        Ok(certs) => certs,
+                        Err(err) => {
+                            error!(%err, "Failed to load certificate");
+                            continue;
+                        }
+                    };
+                    // Load private key.
+                    let key = match load_private_key("/var/run/linkerd/tls/tls.key") {
+                        Ok(key) => key,
+                        Err(err) => {
+                            error!(%err, "Failed to load key");
```

Comment:
similarly, the errors returned by `load_private_key` are prefixed with "failed to load key", so this will duplicate that string in the logged event.